### ISSUE TITLE
Proposal: Optionally store names in ValueTuple, for further processing (databinding)

### ISSUE THREAD

**Issue by vbcodec** [dotnet/roslyn#14395]
If compiler detect function SetItemsNames(String ()) on tuple typy, then should call this function, immediately after creating instance. For example for tuple:

```
    Dim t = (a:="abc".2)`
```
compiler should perform initialization:

```
    Dim o = New ValueTuple(Of String, Integer)("abc", 2)
    If SetNamesExist Then o.SetItemsNames ({"a", ""})
```

Currently, ValueTuple types posted on nuget do not have this function. But some other devs may create such enhanced tuple types, that will contain this function. So, compiler must check for existence of for every tuple type. With stored item names, these enhanced tuple types may implement interface like ICustomTypeDescriptor, that is used by WinForms, WPF, ASP.NET (and others) for databinding. This way usability of tuples will be greatly increased. MS may prepare #such variant of tuple types, to enable devs databinding from tuples,

**Comment by HaloFour** [comment]
Using alternate `ValueTuple` libraries seems like it would be really haphazard for supporting this.  All of a sudden the entire goal of a unified type across assembly would go out the window.  Any framework code written to use a specific custom implementation would be incompatible with any other implementation and will very likely cause reference ambiguities.

I can appreciate the goal of wanting to embed the names into the tuples in a manner that can be extracted from the instances programmatically.  I do think that there will be a relatively unintuitive learning curve regarding where you can use tuples vs. anonymous types, particularly when it comes to reflection-driven APIs (e.g. property bags) or serialization.  I don't think that this is the way to do it, though.

**Comment by vbcodec** [comment]
@HaloFour 

Team chosed that tuples will rely on external types, and that has some consequences. Wit this approach, MS can't guarantee, that users will always use MS library with tuples, especially, that these tuples are very limited (no databinding, and no mutable collections). Combination of those facts, open door do enhance tuples if someone want more than basic support provided by MS. Think about MS tuples as base class, while others may create derived class, if they have needs. For me, I don't want to use tuples as structs, and I've already created my own library with tuples as classes, because I want mutability of tuples in collections. Every code that uses MS tuples will be compatible with custom tuples from other library, but not vice-versa, and nothing is going out of window. Problems will appear only if some dev will want to use two different custom tuples, but this is easy to resolve by keeping custom tuples in separated files. 
So, even without ability to push these names, the doors to potential mess with different libraries, are already opened, and this is accepted by team. I think that team should say something how they perceive these scenarios with custom tuples.

**Comment by HaloFour** [comment]
@vbcodec 

That's true, but the exposure to that is very limited.  You have, what, `ExtensionAttribute`, the built-in family of async method builders (which is being abstracted away) and `INotifyCompleted`?  In most cases that's innocuous as those types are used strictly to enable plumbing and are never a part of a contract.  But tuples may be a part of a contract and to have multiple tuple libraries out there will only introduce confusion.  Imagine two libraries that each reference different tuple libraries, how would a consuming application possibly resolve those ambiguities?  You'd immediately lose the ability to declare a tuple and to pass it between those libraries.  Such equivalence is part of the charter of that feature proposal.  Not to mention, the proposed would enable a specific shape to how names would be tracked without actually providing the functionality.  If name tracking by instance is to be supported that should be done out of the box, not through some hidden method that only third-party libraries *might* support.

Isn't `ValueTuple<...>` slated to be added to the BCL anyway?

**Comment by DavidArno** [comment]
> Isn't `ValueTuple<...>` slated to be added to the BCL anyway?

#13177 seems to suggest so.
